### ISSUE TITLE
Make all SchemaProvider trait APIs async

### ISSUE THREAD

**Issue by metesynnada** [apache/datafusion#10339]
### Is your feature request related to a problem or challenge?

Currently, only `table` API is async, but `register_table`, `deregister_table`, `table_names`, and `table_exist` can be also async. Also, most of the usage is under async methods, thus there shouldn't be a big issue implementing it.

However, this can be a huge change for the downstream usage, this is why I need some insides before me or someone else implements it.

cc @Dandandan @alamb

**Comment by matthewmturner** [comment]
This one is actually quite interesting. I hadnt gotten around to it yet but I was going to propose having both sync and async apis for `SchemaProvider` (I think theres valid use cases for both).  I just recently ran into an issue where I wanted the `ctx.table_provider` and therefore `table` method on `SchemaProvider` to be sync (as part of logical plan serialization which is sync).  How would you feel about having both sync and async APIs?

**Comment by alamb** [comment]
Yes, I agree allowing catalog APIs to be async makes sense to me

One rationale against async, as I recall from @tustvold , was to encourage people to implement efficient catalog implementations (e.g. for many cases calling a network call for `table_exists` might be slow)

However, my personal opinion is that such encouragement can be done via documentation and if people want to implement RPC network calls during planning then the APIs shouldn't stop them

I think the biggest challenge is, as @metesynnada hints at above, the viral nature of `async` -- if we make such APIs async then everywhere they are called must also be be `async` -- I haven't looked at how far down the stack that is but it could be substantail.

An alternate approach might be to implement,  via some hackery and tokio channels,  an struct that implements the `SchemaProvider` without changes (sync) but can call async methods (though that would block the runtime thread 🤔 )

**Comment by tustvold** [comment]
I seem to remember omitting the methods other than table due to the sheer size of the resulting change, I have no objection to making them async

**Comment by westonpace** [comment]
I've made a stab at this.

> However, my personal opinion is that such encouragement can be done via documentation and if people want to implement RPC network calls during planning then the APIs shouldn't stop them

The easiest way we've found to do this kind of thing is with a metadata cache.  However, this cache gets invalidated and has cold start, etc.  The problem with "warming the cache prior to the query" is that it is very difficult to determine which entries will be required by an SQL string.  Loading the entire catalog into memory for a single query is prohibitively expensive for us.

> I think the biggest challenge is, as @metesynnada hints at above, the viral nature of async -- if we make such APIs async then everywhere they are called must also be be async -- I haven't looked at how far down the stack that is but it could be substantail.

Yes :cold_sweat:

> An alternate approach might be to implement, via some hackery and tokio channels, an struct that implements the SchemaProvider without changes (sync) but can call async methods (though that would block the runtime thread 🤔 )

I could find no reasonable way to implement such hackery.
